### PR TITLE
Feature/azure deploy

### PR DESCRIPTION
**Pull request by lmazuel**
##### Issue Type:
- Feature Pull Request
##### Plugin Name:

cloud/azure/azure_deployment
##### Summary:

This is an update of the PR https://github.com/ansible/ansible-modules-extras/pull/1329 using the last Python Azure SDK 2.0.0rc1. The 2.0.0rc1 contains the new Python SDK auto-generated from Swagger and is the best way to deploy in Python for Azure using Azure Resource Management.

Please tell me if I miss something important for the contribution, first contribution here :)

### CODE REVIEW

**Comment by nitzmahone** [comment]
@lmazuel @devigned - Not sure how you guys want to resolve this- looks like lmazuel updated your PR #1329 to use the new 2.0 Azure Python SDK with this one. The pending SDK update was the main reason the core team paused dev on our azure_rm modules (and why we've held off merging this one). Since the 2.0 alpha SDK is out, we've picked the work back up, and are currently planning to ship our modules in Ansible 2.1. Ours are resource-specific modules for resourcegroup, storageaccount, storageblob, virtualnetwork, subnet, publicip, networkinterface, virtualmachine, as well as a new dynamic inventory script and some other bits, so this module is a great generic complement for applying an arbitrary RM template. We definitely want all the Azure RM stuff to target the same SDK version. Once we've merged our stuff into the main Ansible codebase, our shared auth layer will also be available in module_utils, so we'll want to update this with those as well. 

@devigned - looks like @lmazuel based his PR off yours, so you'll both share some "credit" as far as GitHub's concerned. You OK if we close #1329 in favor of this one?

**Comment by devigned** [comment]
I'm going to close mine out in favor of this pr.

**Comment by gregdek** [comment]
After some review, it looks as though this PR needs some revisions (travis failures).

Please make the suggested revisions, and when you're done, please add a comment with the text "ready_for_review" and we will continue with the review process.  Thanks!

**Comment by nitzmahone** [comment]
Ah yep, yay for automated modules docs testing!

**Comment by devigned** [comment]
~~@lmazuel can you fix that errant colon in the docs?~~

**Comment by lmazuel** [comment]
ready_for_review

**Comment by lmazuel** [comment]
Please tell me if I can do everything else to improve the PR. Thank you for your time!

File: cloud/azure/azure_deployment.py [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+    return [dict(vm_name=vm.resource_name, ips=[get_ip_dict(ip) for ip in ips]) for vm, ips in vms_and_ips if len(ips) > 0]
+
+
+def main():
+    argument_spec = dict(
+        azure_url=dict(default=AZURE_URL),
+        subscription_id=dict(),
+        client_secret=dict(no_log=True),
+        client_id=dict(required=True),
+        tenant_id=dict(required=True),
```
Comment (devigned) [inline cloud/azure/azure_deployment.py:565]: client_secret, client_id, tenant_id and anything else that can be supplied in the environment should not be required here. The module should fail after checking either env or provided on the task.

File: cloud/azure/azure_deployment.py [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+                deployment_result = client.deployments.get(group_name, deployment_name)
+                time.sleep(module.params.get('wait_for_deployment_polling_period'))
+
+        if deployment_result.properties.provisioning_state == 'Succeeded':
+            return deployment_result
+
+        failed_deployment_operations = get_failed_deployment_operations(module, client, group_name, deployment_name)
+        module.fail_json(msg='Deployment failed. Deployment id: %s' % (deployment_result.id), failed_deployment_operations=failed_deployment_operations)
+    except CloudError as e:
+        failed_deployment_operations = get_failed_deployment_operations(module, client, group_name, deployment_name)
```
Comment (devigned) [inline cloud/azure/azure_deployment.py:479]: If there is a failure of the deployment (params or template is wrong) there will be no deployment created. A 400 will be generated from line 466, then line 479 will fail with a 404.

I'd expect the 400 failure to provide failure messaging with the reason the 400 was encountered.

File: cloud/azure/azure_deployment.py (hunk 1 above)
Comment (lmazuel) [inline cloud/azure/azure_deployment.py:565]: I stuck to @sedouard behaviour in his PR to be the most homogeneous possible. If you think this does not have to be required, we should discuss with @sedouard too.
I trust you, you know better than me the parameters logic :)

Comment (obsoleted) [inline cloud/azure/azure_deployment.py:565]: I agree with @devigned (the folks we had using this also preferred that method). So please make it optional falling back to the environment (or so is my vote).

Comment (nitzmahone) [inline cloud/azure/azure_deployment.py:565]: I wouldn't worry about it much- that stuff will all change when this gets merged anyway. We've got shared module_utils code for the other Azure RM modules that does all this stuff (as well as defining an on-disk credential storage similar to the way boto/aws work).

**Comment by mscherer** [comment]
I think you should squash the commits in 1 single commit

**Comment by nitzmahone** [comment]
@mscherer FYI- we usually only care about squashed commits for items that are likely to be cherry-picked into a branch other than devel, or at high risk of needing to be backed out.

**Comment by mscherer** [comment]
Well, that's what is asked to some others PR, so I just tried to be consistent, but if that's fine for others, then ok.

**Comment by nitzmahone** [comment]
Yeah, that's definitely something we could clarify in the committer guidelines... :)

**Comment by mscherer** [comment]
I didn't knew we had commiters guidelines, but i was planning to look for it.

**Comment by nitzmahone** [comment]
@mscherer WIP version at https://github.com/ansible/ansible/blob/devel/docsite/rst/committer_guidelines.rst

**Comment by gregdek** [comment]
Thanks @lmazuel for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by chouseknecht** [comment]
Merging this in. Will update to work with common Azure module.

**Pull request merged**